### PR TITLE
Specviz uc bug fix

### LINKED ISSUE #215: Data selection not working properly after unit conversion
After running unit conversion plugin, spectrum viewer data selection switches to copy with converted units, which
is as expected.  If you then select both the converted and unconverted data, only the converted data shows up, which is fine. However, when the converted data is unselected, now two data sets appear (with only the unconverted data selected), which is an unexpected and incorrect behavior.
specviz_unitconversion

### LINKED ISSUE #259: Unit conversion returns error that mentions uncertainty
Traceback:
```
---------------------------------------------------------------------------
UnitConversionError                       Traceback (most recent call last)
~/anaconda3/envs/specviz_uc_bug/lib/python3.8/site-packages/astropy/nddata/nduncertainty.py in parent_nddata(self, value)
    232                 try:
--> 233                     unit_from_data.to(self.unit)
    234                 except UnitConversionError:

~/anaconda3/envs/specviz_uc_bug/lib/python3.8/site-packages/astropy/units/core.py in to(self, other, value, equivalencies)
    986         else:
--> 987             return self._get_converter(other, equivalencies=equivalencies)(value)
    988 

~/anaconda3/envs/specviz_uc_bug/lib/python3.8/site-packages/astropy/units/core.py in _get_converter(self, other, equivalencies)
    917 
--> 918             raise exc
    919 

~/anaconda3/envs/specviz_uc_bug/lib/python3.8/site-packages/astropy/units/core.py in _get_converter(self, other, equivalencies)
    902         try:
--> 903             return self._apply_equivalencies(
    904                 self, other, self._normalize_equivalencies(equivalencies))

~/anaconda3/envs/specviz_uc_bug/lib/python3.8/site-packages/astropy/units/core.py in _apply_equivalencies(self, unit, other, equivalencies)
    885 
--> 886         raise UnitConversionError(
    887             "{} and {} are not convertible".format(

UnitConversionError: 'Jy' (spectral flux density) and 'erg / (Angstrom cm2 s)' (spectral flux density wav) are not convertible

During handling of the above exception, another exception occurred:

UnitConversionError                       Traceback (most recent call last)
~/anaconda3/envs/specviz_uc_bug/lib/python3.8/site-packages/ipyvue/VueTemplateWidget.py in _handle_event(self, _, content, buffers)
     48             event = content.get("event", "")
     49             data = content.get("data", {})
---> 50             getattr(self, 'vue_' + event)(data)
     51 
     52 

~/Documents/jdadf_notebooks/jdaviz/jdaviz/configs/specviz/plugins/unit_conversion/unit_conversion.py in vue_unit_conversion(self, *args, **kwargs)
    143                 return
    144         # Create new spectrum with new units.
--> 145         converted_spec = self.spectrum._copy(flux=set_flux_unit, spectral_axis=set_spectral_axis_unit, unit=set_flux_unit.unit, uncertainty=StdDevUncertainty(self.spectrum.uncertainty))
    146 
    147         # Finds the '_units_copy_' spectrum and does unit conversions in that copy.

~/anaconda3/envs/specviz_uc_bug/lib/python3.8/site-packages/specutils/spectra/spectrum1d.py in _copy(self, **kwargs)
    236         alt_kwargs.update(kwargs)
    237 
--> 238         return self.__class__(**alt_kwargs)
    239 
    240     @property

~/anaconda3/envs/specviz_uc_bug/lib/python3.8/site-packages/specutils/spectra/spectrum1d.py in __init__(self, flux, spectral_axis, wcs, velocity_convention, rest_value, redshift, radial_velocity, **kwargs)
    157                         spectral_axis.shape[0], flux.shape[-1]))
    158 
--> 159         super(Spectrum1D, self).__init__(
    160             data=flux.value if isinstance(flux, u.Quantity) else flux,
    161             wcs=wcs, **kwargs)

~/anaconda3/envs/specviz_uc_bug/lib/python3.8/site-packages/astropy/nddata/nddata.py in __init__(self, data, uncertainty, mask, wcs, meta, unit, copy)
    232         self._unit = unit
    233         # Call the setter for uncertainty to further check the uncertainty
--> 234         self.uncertainty = uncertainty
    235 
    236     def __str__(self):

~/anaconda3/envs/specviz_uc_bug/lib/python3.8/site-packages/astropy/nddata/nddata.py in uncertainty(self, value)
    324                 # to be saved as weakref but that's done by NDUncertainty
    325                 # setter).
--> 326                 value.parent_nddata = self
    327         self._uncertainty = value

~/anaconda3/envs/specviz_uc_bug/lib/python3.8/site-packages/astropy/nddata/nduncertainty.py in parent_nddata(self, value)
    233                     unit_from_data.to(self.unit)
    234                 except UnitConversionError:
--> 235                     raise UnitConversionError("Unit {} of uncertainty "
    236                                               "incompatible with unit {} of "
    237                                               "data".format(self.unit,

UnitConversionError: Unit 1e-17 erg / (Angstrom cm2 s) of uncertainty incompatible with unit Jy of data
```

### PR DESCRIPTION
**Pull request by javerbukh**
Fixes #259 
Fixes #215

### CODE REVIEW

File: jdaviz/configs/specviz/plugins/unit_conversion/unit_conversion.py [hunk 1]
```diff
@@ -140,8 +141,18 @@ def vue_unit_conversion(self, *args, **kwargs):
...
+                                                      equivalencies=u.spectral_density(set_spectral_axis_unit))
+
         # Create new spectrum with new units.
-        converted_spec = self.spectrum._copy(flux=set_flux_unit, spectral_axis=set_spectral_axis_unit, unit=set_flux_unit.unit)
+        converted_spec = self.spectrum._copy(flux=set_flux_unit,
+                                             spectral_axis=set_spectral_axis_unit,
+                                             unit=set_flux_unit.unit,
+                                             uncertainty=self.return_uncertainty_type(self.spectrum.uncertainty.uncertainty_type,
+                                                                                      temp_uncertainty.value)
+                                             )
```
Comment (nmearl) [inline jdaviz/configs/specviz/plugins/unit_conversion/unit_conversion.py:155]:
```suggestion
                                             uncertainty=self.spectrum.uncertainty.__class__(temp_uncertainty.value))
```

I don't think you need to create a method for dealing with uncertainty types, instead just use the class of the input uncertainties. This has the benefit of seamlessly handling any new types of uncertainties the user is using that aren't specified in your `if`/`else` statement.

File: jdaviz/configs/specviz/plugins/unit_conversion/unit_conversion.py [hunk 2]
```diff
@@ -277,3 +288,28 @@ def convert_units_to_strings(self, unit_list):
...
+            Uncertainty type of original spectrum with the newly converted values for uncertainty.
+        """
+        if uncertainty_type == "std":
+            return StdDevUncertainty(value)
+        elif uncertainty_type == "var":
+            return VarianceUncertainty(value)
+        elif uncertainty_type == "ivar":
+            return InverseVariance(value)
+        else:
+            return UnknownUncertainty(value)
```
Comment (nmearl) [inline jdaviz/configs/specviz/plugins/unit_conversion/unit_conversion.py:315]:
```suggestion
```

File: jdaviz/configs/specviz/plugins/unit_conversion/unit_conversion.py [hunk 3]
```diff
@@ -140,8 +141,18 @@ def vue_unit_conversion(self, *args, **kwargs):
                 self.hub.broadcast(snackbar_message)
 
                 return
+
+        # Uncertainty converted to new flux units
+        temp_uncertainty = self.spectrum.uncertainty.quantity.to(u.Unit(set_flux_unit.unit),
+                                                      equivalencies=u.spectral_density(set_spectral_axis_unit))
```
Comment (nmearl) [inline jdaviz/configs/specviz/plugins/unit_conversion/unit_conversion.py:147]:
```suggestion
                                                      equivalencies=u.spectral_density(set_spectral_axis_unit))
```

Does this actually work? `u.spectral_density` should be taking an array of dispersion values, not a unit object...

**Comment by dtn5ah** [comment]
Ack, I was able to reproduce #260 on this branch. Using Specviz, converting https://data.sdss.org/sas/dr14/sdss/spectro/redux/26/spectra/0751/spec-0751-52251-0160.fits to micron/microJansky

image

File: jdaviz/configs/specviz/plugins/unit_conversion/unit_conversion.py (hunk 1 above)
Comment (javerbukh) [reply jdaviz/configs/specviz/plugins/unit_conversion/unit_conversion.py:155]: I was looking for some way to do this but couldn't find anything in the documentation, thanks!

File: jdaviz/configs/specviz/plugins/unit_conversion/unit_conversion.py (hunk 3 above)
Comment (javerbukh) [reply jdaviz/configs/specviz/plugins/unit_conversion/unit_conversion.py:147]: `set_spectral_axis_unit` is actually the converted spectral axis, not just the unit.

**Comment by javerbukh** [comment]
@dtn5ah I'm not seeing that error anymore. Did you do `pip install -e .` after pulling down the branch?

**Comment by rosteen** [comment]
FYI I could reproduce Duy's error in an existing env that I tried pulling and installing, but running off a new install in a clean env works without errors. Likely a conflict with an older version of glue-jupyter, if I had to guess. @dtn5ah did you see this running in a newly created environment?

**Comment by dtn5ah** [comment]
@rosteen Yep it was a clean env. Tried it twice, once with `pip install git+https://github.com/spacetelescope/jdaviz.git@refs/pull/260/head` and a second, after @javerbukh's recommendation `pip install -e git+https://github.com/spacetelescope/jdaviz.git@refs/pull/260/head#egg=jdaviz`

**Comment by rosteen** [comment]
Strange...I don't see why it would be any different, but perhaps try pulling the repo down with git and installing from the local copy, just in case?
```
git checkout -b javerbukh-specviz_uc_bug_fix master
git pull https://github.com/javerbukh/jdaviz.git specviz_uc_bug_fix
```

**Pull request merged**